rocAL rocJpeg Integration

fiona-gladwin opened the pull request:
- Introduce rocJpeg hardware decoder in rocAL

fiona-gladwin committed:
Add rocJpeg files initial commit

fiona-gladwin committed:
Add support for RocJpeg decoder in rocAL

Add changes in CMake

fiona-gladwin committed:
Fix issues with rocJpeg decoder

fiona-gladwin committed:
Remove additional copy to circular buffer

fiona-gladwin committed:
Adjust pitch to max width and height

fiona-gladwin committed:
Add batch decoding support

fiona-gladwin committed:
Add Resize HIP kernel for rocJpeg decode

fiona-gladwin committed:
Add support for scaling similar to TurboJpeg in rocJpeg decoder

fiona-gladwin committed:
Add support to store src image contiguously

fiona-gladwin committed:
Add generic HIP kernel support

fiona-gladwin committed:
rocJpeg : Add support for Grayscale images decoding

fiona-gladwin committed:
Fix resize output corruption

fiona-gladwin committed:
Pass HIP stream from device resouces to rocjpeg decoder

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Add support to eliminate resize when images are not scaled

fiona-gladwin committed:
Eliminate failing images

fiona-gladwin committed:
Fix build issues and seg fault

fiona-gladwin committed:
Add ChromaSubsampling to decode info

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/rocjpeg_integration2

fiona-gladwin committed:
Reduce number of rocJpeg API calls

fiona-gladwin committed:
Add support to handle failing images in the batch

fiona-gladwin committed:
Unit test change

fiona-gladwin committed:
Fix warnings

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Fix failing images

fiona-gladwin committed:
Minor fix

fiona-gladwin committed:
Reintroduce decode info

Comment out print statements

fiona-gladwin committed:
Add support to use HIP memory in circular buffer for HW decoder

fiona-gladwin committed:
Change to rocjpeg hw decoder in API

fiona-gladwin committed:
Remove decode info batch call

fiona-gladwin committed:
Overload initialize and decode_info call for hardware decode

fiona-gladwin committed:
Add API calls to release rocJpeg resources

fiona-gladwin committed:
Remove unsupported check for subsampling 440

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Fix segmentation fault

Change the name of the RocJpegDecoder class

fiona-gladwin committed:
Set device ID for each thread

fiona-gladwin committed:
Reset device ID for each thread

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Introduce rocJpeg cmake and update changes

fiona-gladwin committed:
Introduce rocJpeg decoder in api types

fiona-gladwin committed:
Code clean up and add comments

fiona-gladwin committed:
Update decode info and batch API with doc

Remove unused args in decode_batch API

fiona-gladwin commented on rocAL/rocAL_hip/rocal_hip_kernels.cpp:
Remove unused API and functions added

fiona-gladwin commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
Check if any of the private variables are unused

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Move set device ID to image read and decode

Remove reset device ID API

fiona-gladwin committed:
Revert unit test changes

SundarRajan28 committed:
Adding rocjpeg decoder type in pybind

shobana-mcw commented on rocAL/include/decoders/image/decoder.h:
```diff
@@ -89,6 +103,24 @@ class Decoder {
                                int *color_comps) = 0;
+    
+    //! Decodes the header of the Jpeg compressed data and returns basic info about the compressed image
+    //! It also scales the width and height wrt max decoded width and height
```
Does rocJPEG scales the width and height? Please check the comment.

fiona-gladwin replied:
Yes, in the rocJpeg decoder we do the scaling in accordance with the max decoded width and height

shobana-mcw commented on rocAL/include/decoders/image/decoder.h:
```diff
@@ -106,8 +138,24 @@ class Decoder {
+
     virtual ~Decoder() = default;
     virtual void initialize(int device_id) = 0;
+    virtual void initialize(int device_id, unsigned batch_size) { THROW("Initialize not implemented") }
```
Why is this not implemented?

fiona-gladwin replied:
This initialize overriden function is used only in rocJpeg decoder where we need to pass batch size as well for other decoders this function will not be defined hence added the THROW statement

shobana-mcw commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,139 @@
+/*
+Copyright (c) 2024 Advanced Micro Devices, Inc. All rights reserved.
```
Change the copyrights Year.

shobana-mcw commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,139 @@
+#endif
+
+    unsigned _num_scaling_factors = 16;
+    ScalingFactor _scaling_factors[16] = {
```
Check if the scaling factor is required. Also please check for unused variables or functions and remove them.

fiona-gladwin replied:
In rocJpeg decoder we do the scaling similar to how TurboJpeg does hence scaling factor is required

SundarRajan28 replied:
Instead of hardcoding, better to use tjGetScalingFactors API to fill the scaling factors. Refer [this](https://github.com/ROCm/rocAL/blob/develop/rocAL/source/decoders/image/turbo_jpeg_decoder.cpp#L31)

fiona-gladwin replied:
That would require including the turboJpeg lib inside the rocjpeg_decoder.cpp, is that fine @shobana-mcw

shobana-mcw commented on rocAL/rocAL_hip/rocal_hip_kernels.cpp:
```diff
@@ -296,3 +307,453 @@ int HipExecCopyInt8ToNCHW(
     return 0;
 }
+
+__device__ void resize_roi_and_srclocs_hip_compute(int4 *srcRoiPtr_i4, uint2 *dstDimsWH, int id_x, int id_y, d_float16 *locSrc_f16)
```
Change the variable names wrt rocAL naming conventions.

fiona-gladwin replied:
Done

shobana-mcw commented on rocAL/source/decoders/image/rocjpeg_decoder.cpp:
```diff
@@ -0,0 +1,386 @@
+#define CHECK_ROCJPEG(call) {                                             \
+    RocJpegStatus rocjpeg_status = (call);                                \
+    if (rocjpeg_status != ROCJPEG_STATUS_SUCCESS) {                       \
+        std::cerr << #call << " returned " << rocJpegGetErrorName(rocjpeg_status) << " at " <<  __FILE__ << ":" << __LINE__ << std::endl;\
```
Should we use our error statements?

fiona-gladwin replied:
Should I use THROW here?

shobana-mcw commented on rocAL/source/loaders/image/image_read_and_decode.cpp:
```diff
@@ -279,54 +289,109 @@ ImageReadAndDecode::load(unsigned char *buff,
+                        } else
+                            j--;
+                        if (j < 0) {
+                            THROW("All images in the batch failed decoding\n");
```
Add more info to this error statement mentioning the decoder used.

fiona-gladwin replied:
Done

shobana-mcw commented on rocAL/source/loaders/circular_buffer.cpp:
```diff
@@ -151,7 +155,8 @@ void CircularBuffer::pop() {
 }
-void CircularBuffer::init(RocalMemType output_mem_type, size_t output_mem_size, size_t buffer_depth) {
+void CircularBuffer::init(RocalMemType output_mem_type, size_t output_mem_size, size_t buffer_depth, bool use_hip_memory) {
+    _use_pinned_memory = !use_hip_memory; // When using Hardware decoder, pinned memory is not allocated for HIP backend
```
Are we still allowing HIP backend where rocJPEG is not used for decoder?

fiona-gladwin replied:
We have a condition check in image_loader.cpp
When the rocJpeg decoder is enabled we set this use_hip_memory to true, else it will be true

fiona-gladwin committed:
Merge branch 'develop' of https://github.com/ROCm/rocAL into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Add missing audio generic decoder type in pybind

fiona-gladwin committed:
Change copyright year and minor changes

SundarRajan28 commented on rocAL/CMakeLists.txt:
```diff
@@ -389,6 +391,14 @@ if(${BUILD_ROCAL})
     endif()
+
+    # rocJPEG
+    if("${BACKEND}" STREQUAL "HIP" AND HIP_FOUND AND ROCJPEG_FOUND)
```
@fiona-gladwin Is rocJpeg HW decoder an optional feature or required feature? If its optional, can we add a CMake flag to enable HW decoder? I can only see #ENABLE_HIP guards around rocJpeg decoder functions. Does that mean rocJpeg is required to be installed always if HIP is present?

SundarRajan28 replied:
Better to add a CMake flag like ENABLE_ROCJPEG_DECODE

fiona-gladwin replied:
Done

SundarRajan28 commented on rocAL/include/decoders/image/decoder.h:
```diff
@@ -37,7 +43,8 @@ enum class DecoderType {
     ROCDEC_VIDEO_DECODE = 7, //!< for video decoding using HW via rocDecode
-    AUDIO_SOFTWARE_DECODE = 8   //!< Uses sndfile to decode audio files
+    AUDIO_SOFTWARE_DECODE = 8,  //!< Uses sndfile to decode audio files
+    ROCJPEG_DEC = 9             //!< rocJpeg hardware decoder for decoding jpeg files
```
Rename as ROCJPEG_DECODE

SundarRajan28 commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,139 @@
+
+#pragma once
+#include "decoders/image/decoder.h"
+#include "rocjpeg.h"
```
Move rocjpeg include to L24.

fiona-gladwin replied:
Moved inside the ENABLE_ROCJPEG portion

SundarRajan28 commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,139 @@
+    //! Default constructor
+    HWRocJpegDecoder();
+#if ENABLE_HIP
+    HWRocJpegDecoder(hipStream_t &stream) { _hip_stream = stream; }
```
Is there an option to use HWRocJpegDecoder without HIP? If not, the entire class definition should be within ENABLE_HIP if guard. Better to use ENABLE_ROCJPEG_DECODE flag

fiona-gladwin replied:
Done

SundarRajan28 commented on rocAL/include/decoders/image/rocjpeg_decoder.h:
```diff
@@ -0,0 +1,139 @@
+    */
+    Status decode_info(unsigned char *input_buffer, size_t input_size, int *width, int *height, int *actual_width, int *actual_height, int max_decoded_width, int max_decoded_height, Decoder::ColorFormat desired_decoded_color_format, int index) override;
+
+    Decoder::Status decode(unsigned char *input_buffer, size_t input_size, unsigned char *output_buffer,
```
Mention in all instances if any of the input_buffer is required as a HIP buffer. Mention like
"User provided HIP buffer ..."

SundarRajan28 commented on rocAL/source/loaders/image/image_loader.cpp:
```diff
@@ -38,6 +38,10 @@ ImageLoader::ImageLoader(void *dev_resources) : _circ_buff(dev_resources),
     _remaining_image_count = 0;
     _device_id = 0;
+#if ENABLE_HIP
+    DeviceResourcesHip *hipres = static_cast<DeviceResourcesHip *>(dev_resources);
```
Is there any other way we can use existing HIP stream from circular buffer or master_graph? We already create 2 HIP streams in rocAL before this and RPP creates one for HIP processing. Will creating multiple HIP streams reduce HIP performance?

fiona-gladwin replied:
We are not creating new stream here, we are using the existing stream [here ](https://github.com/fiona-gladwin/rocAL/blob/af0a0be4faf67a74339020769b38470c7ff1359f/rocAL/source/loaders/image/image_loader.cpp#L148)that is used in circular buffer

SundarRajan28 commented on rocAL/source/loaders/image/image_read_and_decode.cpp:
```diff
@@ -279,54 +289,109 @@ ImageReadAndDecode::load(unsigned char *buff,
-            _actual_decoded_width[i] = scaledw;
-            _actual_decoded_height[i] = scaledh;
         }
+
```
Reduce to 2 line breaks

fiona-gladwin committed:
Use snake case for variables in HIP kernels

fiona-gladwin committed:
Add ENABLE ROCJPEG flag

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Add comment

fiona-gladwin committed:
Update CHANGELOG and Readme

fiona-gladwin committed:
Minor change

kiritigowda committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Add CSS_440 support

fiona-gladwin committed:
Minor changes

fiona-gladwin committed:
Merge branch 'fg/rocjpeg_integration2.1' of https://github.com/fiona-gladwin/rocAL into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

fiona-gladwin committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Minor change

kiritigowda committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

kiritigowda committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

kiritigowda committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Convert JpegDecode params to a vector

fiona-gladwin committed:
Add ROI check in GetChannelPitchAndSizes

lucbruni-amd committed:
Test - Fix README inconsistencies (#299)

* Test - Fix README inconsistencies

* Test - Fix typo

* Update tests/cpp_api/unit_tests/README.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

---------

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

dependabot[bot] committed:
Docs - Bump rocm-docs-core[api_reference] from 1.17.1 to 1.18.1 in /docs/sphinx (#300)

Bumps [rocm-docs-core[api_reference]](https://github.com/ROCm/rocm-docs-core) from 1.17.1 to 1.18.1.
- [Release notes](https://github.com/ROCm/rocm-docs-core/releases)
- [Changelog](https://github.com/ROCm/rocm-docs-core/blob/develop/CHANGELOG.md)
- [Commits](https://github.com/ROCm/rocm-docs-core/compare/v1.17.1...v1.18.1)

---
updated-dependencies:
- dependency-name: rocm-docs-core[api_reference]
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

fiona-gladwin committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

SundarRajan28 committed:
Deps update: Support turbojpeg package (#296)

* Removing tj3 API usage

* Adding turbojpeg package in rocAL setup script

* Update changelog

* Add turbojpeg rpm pkg

---------

Co-authored-by: SundarRajan98 <svaithiy@amd.com>
Co-authored-by: Lakshmi Kumar <lakshmi.kumar@amd.com>

fiona-gladwin committed:
Update CMake to find rocJpeg

fiona-gladwin committed:
Add rocJpeg version check in CMakeLists

fiona-gladwin committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

kiritigowda committed:
Package - TurboJPEG deps (#303)

* Package - TurboJPEG deps

* Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

* Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

* SLES - Updates for TurboJPEG

* CMakeLists - Updates

* Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

* Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

* Update CHANGELOG.md

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

* Cleanup

---------

Co-authored-by: spolifroni-amd <Sandra.Polifroni@amd.com>

kiritigowda committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

SundarRajan28 committed:
Numpy reader - C++ changes (#291)

* Fix shard_size and audio source evaluation

* Changes in file_source_reader - to minimize the I/O operations

* Changes in the variable name

* Changes in the variable names of the audio source evalution

* Use set instead of vector

* Minor bug fixes

* Minor fixes

Remove changes to update the
filenames vector incase of padding
Fix the pipeline when pad is off and use idx in a continuous manner

* Fix drop policy without padding

To skip and batch and start with the previously padded idx in last batch

* Fix pytorch iterator - PARTIAL policy

* Revert empty line removed in CMakeLists.txt

* Removed prefix original for audio vectors

* Fix PARTIAL

Fix issue with batch size greater than dataset size

* Reduce overall time for audio source evalution

* Fix shard_size and stick to shard issue seen with convergence

* Resolve PR comments

* Add @params to all args in pytorch.py

* Fix build issue

* Minor changes in unit test

* Minor changes

* Change ROCAL instaces to rocAL in pytorch.py

* Resolve the PR comments

* Minor changes in decoders.py - Modify the comment for shard_size

* Fix shard_size

* Minor changes

* Changes in pipeline.py and decoders.py

* Address the PR comments

* Address Review comments

* Remove print statement

* Fix the count_items

* Make Sharding similar to DALI

* Fix issues with DROP policy by introducing a new vector for padding

* Minor fixes

* Comment out print statements

* Add changes for shard_size LBP testing

* Fix DROP Policy with shard_size > 0

* Fix Stick_to_Shard=False

* Fix PARTIAL policy and code clean up

* fix last_batch_padded size when shard_size > 0

* Fix Drop policy - we skip the dropped batch in the next epoch

* Fix single shard outputs

* Remove the commented code and fix the padding code in open()

* Remove div by num_shards in decoders.py

* Introduce Audio layouts

* Add layout changes for spectrogram

* Fix the unit tests - c++ & python

* Code clean up and formatting

* Minor code clean up

* code clean up in pytorch.py

* Add layout changes for spectrogram

* Pass layouts for MelFilterBank

* Fix ToDecibels

Pass layouts for ToDecibels

* Fix Normalize

* Fix build issue

* Fix python unit test

* Minor fix

* Pass LBP to decoders instead of the Pipeline creation

* Update pipeline.py - Remove commented code

* Update pipeline.py - Remove commented out code

* Adding changes for spec layout changes

* Adding changes to MFB and normalize nodes

* Update node_slice.cpp

* Update node_slice.h

* Resolve PR comments

* Fix downmix failing case and resolve the issue with merge

* Fix issue with file_source_reader.cpp when file_list is not used

* Add the changes for LBP in coco-file-source-reader.

* Resolve PR comments - Sundar

* Fix undefined reference

* Fix file_source_reader.cpp

* Fix shuffle issues

* Fix coco_file_source

* make stick_to_shard as true by default

* Fix minor bug

* Change last_patch_padded to pad_last_batch_repeated

* Fix the shuffle with aspect ratios

* Make a minor change in the coco-file-source - sorting is not required

* Another fix for shuffle with aspect_ratios

* Another fix for shuffle with aspect_ratios

* Modify the split_count

* Fix name of shard_count

* Adding print statements

* Add the sorting back

* Fix the indexs calc

* Fix the indexs calc

* Add some debug statements

* Fix the ARG

* comment out part of ASR

* Comment out shuffle_with_asr

* Uncomment Line 234 & 236

* Uncomment Line 234 & 236

* Uncomment the ARG

* Add Cmake changes for HIP backend

* Add support for TF record reader

* Adding changes for lbp & sharding in caffe reader

* Adding changes for lbp & sharding in caffe2 lmdb reader

* Adding changes for lbp & sharding in cifar10 reader

* Adding changes for lbp & sharding in mxnetIO reader

* Minor change in tf record reader

* Adding partial LBP changes to np reader

* Adding comments to all if conditions

* Adding the lbps function implementation

* Minor change

* Fix a bug in the caffe2 reader

* Minor changes

* Fix count_items

* Fix merge conflicts

* Resolving review comments

* Resolving review comments

* Fix a minor warning in file source reader

* Resolving review comments

* LBP comments resolution

* Resolving review comments

* Formatting changes

* Resolving Final Set of PR comments

* Combine with OR condition

* Remove the pad_last_batch_repeated print statement from decoders.py

* Add shard_size and stick_to_shard variables in args

* Minor spelling fix

* Make changes to insert the padded data in the file_names vector

* Support to pass the variables fo lbp as struct

* Fix segmentation fault

* Removing PR11 changes

* Adding LBP changes to numpy reader

* Adding i16 support to rocAL and np reader

* Add npy version check

* Resolving review comments

* Adding numpy loader changes

* Change copyright year

* Resolving review comments

* Resolve review comments

* Fixing issue with setting dtype

* Minor changes

* Adding ROI changes for numpy reader

* Updating unit test for numpy reader

* Resolving review comments

* Adding function comments

* Adding checks for numpy header parsing

* Adding fixes for checks

* Adding changes from PR267

* Resolving review comments

* Resolving review comments

* Resolving review comments

* Macro changes

* Removing templated function for reading numpy data

* Resolving review comments

* Minor changes

---------

Co-authored-by: fiona-gladwin <fionagladwin@multicorewareinc.com>
Co-authored-by: Swetha B S <swetha@multicorewareinc.om>
Co-authored-by: Swwetha B S <swetha@multicorewareinc.com>
Co-authored-by: root <root@x1000c4s5b0n0.hostmgmt2000.cm.lockhart.amd.com>
Co-authored-by: swetha097 <59434434+swetha097@users.noreply.github.com>
Co-authored-by: SundarRajan98 <svaithiy@amd.com>
Co-authored-by: sbavasab <102275787+sbavasab@users.noreply.github.com>
Co-authored-by: Lakshmi Kumar <lakshmi.kumar@amd.com>
Co-authored-by: Kiriti Gowda <kiritigowda@gmail.com>

fiona-gladwin committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Update the CMakeLists and setup script with rocJpeg packages

fiona-gladwin committed:
Update docs

fiona-gladwin committed:
Update basic test with decoder type argument

fiona-gladwin committed:
Add ctest for rocJpeg integration

fiona-gladwin committed:
Remove comment

LakshmiKumar23 committed:
rocDecode Integration: adding ctest  (#304)

fiona-gladwin committed:
Merge branch 'develop' into fg/rocjpeg_integration2.1

fiona-gladwin committed:
Update Ctest for rocJpeg to detect rocJpeg instead of HIP
